### PR TITLE
Geosubscriptions

### PR DESCRIPTION
**Pull request by fgalan**
Implements #1678

### CODE REVIEW

File: src/lib/mongoBackend/MongoGlobal.cpp [hunk 1]
```diff
@@ -835,14 +835,68 @@ static bool processAreaScope(const Scope* scoP, BSONObj &areaQuery)
 }
 
 
+
+
+/* *****************************************************************************
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Not an actual change in addFilterScope() and sortCriteria(). Actually, I think that I dindn't move them, what I do was to remove `static` in processAreaScopeV2()  and move that function (but not the others).

However, the diff is shown in this weird way...

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 2]
```diff
@@ -1576,8 +1576,58 @@ static bool processSubscriptions
...
+      }
+
+      BSONObj areaFilter;
+      if (!processAreaScopeV2(&geoScope, areaFilter))
+      {
+        // Error in processAreaScopeV2 is intepreted as no-match (conservative approach)
+        continue;
+      }
+
+      // Look in the database of and entity that maches the geo-filters. Note that this query doesn't
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: of and ...

for an ?

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 3]
```diff
@@ -1576,8 +1576,58 @@ static bool processSubscriptions
...
+      // check any other filtering condition, asumming they are already checked in other steps.
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: typo: assuming

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 4]
```diff
@@ -1576,8 +1576,58 @@ static bool processSubscriptions
...
+      std::string  id      = notifyCerP->contextElement.entityId.id;
+      std::string  type    = notifyCerP->contextElement.entityId.type;
+      std::string  sp      = notifyCerP->contextElement.entityId.servicePath;
+
+      BSONObj query = BSON(keyId << id << keyType << type << keySp << sp << keyLoc << areaFilter);
+
+      unsigned long long n;
+      if (!collectionCount(getEntitiesCollectionName(tenant), query, &n, &filterErr))
+      {
+        // Error in database access is intepreted as no-match (conservative approach)
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: typo: interpreted

File: test/functionalTest/cases/1658_v2_patch_subscription/patch_subscriptions.test [hunk 5]
```diff
@@ -460,19 +460,24 @@ Date: REGEX(.*)
 }
 
 
-09. PATCH /v2/subscriptions (expresssion)
-=========================================
-HTTP/1.1 204 No Content
-Content-Length: 0
+09. PATCH /v2/subscriptions (invalid expresssion)
```
Comment (kzangeli) [inline test/functionalTest/cases/1658_v2_patch_subscription/patch_subscriptions.test]: sss

File: test/functionalTest/cases/1678_geosubscriptions/geosubscriptions.test [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+accumulatorStart --pretty-print
+
+--SHELL--
+
+#
+# 01. Create entity E1 insise area A
```
Comment (kzangeli) [inline test/functionalTest/cases/1678_geosubscriptions/geosubscriptions.test]: inside

File: src/lib/jsonParseV2/parseSubscription.cpp [hunk 7]
```diff
@@ -785,6 +785,21 @@ static std::string parseNotifyConditionVector(ConnectionInfo* ciP, ngsiv2::Subsc
         subsP->subject.condition.expression.georel = georelOpt.value;
       }
     }
+
+    // If geometry, coords and georel are filled, then attemp to create a filter scope
```
Comment (kzangeli) [inline src/lib/jsonParseV2/parseSubscription.cpp]: typo: attempt

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 8]
```diff
@@ -1576,8 +1576,58 @@ static bool processSubscriptions
...
+      Scope        geoScope;
+      std::string  filterErr;
+
+      if (geoScope.fill("v2", tSubP->expression.geometry, tSubP->expression.coords, tSubP->expression.georel, &filterErr) != 0)
+      {
+        // This has been already checked at subscription creation/update parsing time. Thus, the code cannot reach
+        // this part.
+        //
+        // (Probably the whole if clause will disapear when the missing part of #1705 gets implemented,
+        // moving geo-suff strings to a filter object in TriggeredSubscription class
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: typo: geo-stuff

File: test/functionalTest/cases/1316_create_subscription/post_subs_errors.test [hunk 9]
```diff
@@ -225,7 +225,7 @@ payload='
                "q": "temperature>40",
                "geometry": "point",
                "coords": "-40.4,-3.5",
-               "georel": "near;minDistance=5000"
+               "georel": "near;minDistance:5000"
```
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_errors.test]: Did this PR change the '=' for a ':' ?
Or, are all these old errors ?

Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_errors.test]: They are old.

Until now, expression.(georel|coords|geometry) was just blindly stored/retrieved in/from DB without actually parsing them. Thus, when we change `=` to `:` we didn't notice that some .test need an adjustment.

NTC

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 1 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: NTC (just informative)

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 2 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Yes, it is 'an'. I tend to use 'and' in place where I want to use 'an'. Like nofity ;)

Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in 6a25b851c2eee240a24581294abd6ce500adfcd1

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in 6a25b851c2eee240a24581294abd6ce500adfcd1

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 4 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in 6a25b851c2eee240a24581294abd6ce500adfcd1

File: test/functionalTest/cases/1658_v2_patch_subscription/patch_subscriptions.test (hunk 5 above)
Comment (fgalan) [inline test/functionalTest/cases/1658_v2_patch_subscription/patch_subscriptions.test]: Fixed in 6a25b851c2eee240a24581294abd6ce500adfcd1

File: test/functionalTest/cases/1678_geosubscriptions/geosubscriptions.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/1678_geosubscriptions/geosubscriptions.test]: Fixed in 6a25b851c2eee240a24581294abd6ce500adfcd1

File: src/lib/jsonParseV2/parseSubscription.cpp (hunk 7 above)
Comment (fgalan) [inline src/lib/jsonParseV2/parseSubscription.cpp]: Fixed in 6a25b851c2eee240a24581294abd6ce500adfcd1

Not only in this file, but also in a couple of places that "replace all" has found.

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 8 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in 6a25b851c2eee240a24581294abd6ce500adfcd1

**Comment by kzangeli** [comment]
LGTM

**Comment by crbrox** [comment]
LGTM

**Pull request merged**